prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1155 @@
+/*
+ * Copyright (c) Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+
+#include <ofi_enosys.h>
+#include "ofi_iov.h"
+#include <ofi_util.h>
+
+static struct util_rx_entry *util_alloc_rx_entry(struct util_srx_ctx *srx)
+{
+	return (struct util_rx_entry *) ofi_buf_alloc(srx->rx_pool);
+}
+
+static inline struct iovec *util_srx_iov(struct util_rx_entry *rx_entry)
+{
+	return (struct iovec *) ((char *) rx_entry + sizeof(*rx_entry));
+}
+
+static inline void **util_srx_desc(struct util_srx_ctx *srx,
+				   struct util_rx_entry *rx_entry)
+{
+	return (void **) ((char *) util_srx_iov(rx_entry) +
+			(sizeof(struct iovec) * srx->iov_limit));
+}
+
+static void util_init_rx_entry(struct util_rx_entry *entry,
+			       const struct iovec *iov, void **desc,
+			       size_t count, fi_addr_t addr, void *context,
+			       uint64_t tag, uint64_t flags)
+{
+	memcpy(entry->peer_entry.iov, iov, sizeof(*iov) * count);
+	if (desc)
+		memcpy(entry->peer_entry.desc, desc, sizeof(*desc) * count);
+	else
+		memset(entry->peer_entry.desc, 0, sizeof(*desc) * count);
```

Comment:
I kept the allocation because it could be that the owner provider does not need memory registration but its peer does so this keeps the memory there in case it uses a peer that does need it and needs to update the descriptors for the peer.